
# Bulk RNA-seq deconvolution and performance evaluation by scRNA-seq dataset
## Author: Peng Xu
## Date: 11/20/2021
Note: helper_functions codes were adopted from "https://github.com/favilaco/deconv_benchmark/Master_deconvolution.R". Nature Communications; https://doi.org/10.1038/s41467-020-19015-1.
```{r}
source('./scripts/R_functions/CIBERSORT.R')
source('./scripts/R_functions/helper_functions.R')
library(Matrix)
library(Biobase)
library(dplyr)
library(tidyr)
library("EnsDb.Hsapiens.v86")
```

## Input dataset from Cluster
bulk: bulk RNA-seq count matrix; filtered.counts: snRNA-seq count matrix; filtered.colMetadata: meta file for snRNA-seq; markers 
```{r}
bulk = readRDS("counts.geneID.sinai_pipeline.RDS")
filtered.counts <- readMM("filtered_count_matrix.mtx")
rownames(filtered.counts) <- readLines("filtered_gene_row_names.txt")
filtered.colMetadata <- read.delim("filtered_column_metadata.txt")
markers_symbol = read.delim("markers_symbol.tsv")
```
##Preprocessing data matrix
```{r}
bulk_exp = exprs(bulk)
meta = phenoData(bulk)@data
colnames(bulk_exp) = meta[colnames(bulk_exp),"projid"]
row.names(filtered.colMetadata) = filtered.colMetadata[,1]
colnames(filtered.counts) = row.names(filtered.colMetadata)
```

Convert Ensemble names to gene symbols
```{r}
G_list <- ensembldb::select(EnsDb.Hsapiens.v86, keys= row.names(bulk_exp), keytype = "GENEID", columns = c("GENEID","SYMBOL"))
colnames(G_list) = c("gene","Symbol")
row.names(G_list) = G_list$gene
common_gene = as.character(G_list[G_list$Symbol %in% rownames(filtered.counts),"Symbol"])
common_gene_uniq = names(table(common_gene))[table(common_gene)==1]
common_gene_uniq_id = as.character(G_list[G_list$Symbol %in% common_gene_uniq,"gene"]) 
```
Select bulk RNA-seq and snRNA-seq from the same individuals
```{r} 
T_orig = bulk_exp[common_gene_uniq_id,]
row.names(T_orig) = G_list[common_gene_uniq_id,"Symbol"]
common_projid = colnames(T_orig)[colnames(T_orig) %in% unique(filtered.colMetadata$projid)]
T_orig = T_orig[,common_projid]
```

## Build training set
```{r}
set.seed(1)
sample_subset = sample(common_projid,20)
train_cellID = filtered.counts[,filtered.colMetadata$projid %in% sample_subset]
train = train_cellID
```
Construct phenotype dataset pDataC
```{r}
pDataC = filtered.colMetadata[filtered.colMetadata$projid %in% common_projid,]
pDataC = data.frame(cellID = pDataC$TAG, cellType =  pDataC$broad.cell.type, sampleID = pDataC$projid)
pDataC[pDataC$cellType=="Ex" | pDataC$cellType=="In", ]$cellType = "Neu"
row.names(pDataC) = pDataC$cellID
colnames(train) = pDataC[colnames(train),]$cellType
```

## Generate P matrix of cell proportions based on single cell
```{r}
individual = unique(pDataC$sampleID)
individual_fraction = list()
for(i in 1:length(individual)){
  individual_fraction[[i]] = data.frame(individual[i],table(pDataC[pDataC$sampleID == individual[i],"cellType"])/nrow(pDataC[pDataC$sampleID == individual[i],]))
}
P = do.call(rbind.data.frame, individual_fraction)
colnames(P) = c("tissue", "CT", "expected_values")
```
## Generate C matrix of cell-type expressions as the reference for bulk RNAseq based methods
```{r}
normalization = "TMM"
train_TMM = Scaling(train, normalization)
cellType <- colnames(train_TMM)
group = list()
for(i in unique(cellType)){ 
  group[[i]] <- which(cellType %in% i)
}
C_orig = lapply(group,function(x) Matrix::rowMeans(train_TMM[,x])) #C should be made with the mean (not sum) to agree with the way markers were selected
C = do.call(cbind.data.frame, C_orig)
```

## Generate variation matrix of cell-type expressions as the reference for bulk RNAseq based methods
```{r}
refProfiles.var = lapply(group,function(x) train_TMM[,x])
refProfiles.var = lapply(refProfiles.var, function(x) matrixStats::rowSds(Matrix::as.matrix(x)))
refProfiles.var = do.call(cbind.data.frame, refProfiles.var)
rownames(refProfiles.var) <- rownames(C_orig)
```

## Transformation, scaling/normalization, and marker selection
```{r}
T = Scaling(T_orig, normalization)
train_cellID_TMM =  Scaling(train_cellID, normalization)
marker_strategy = "all"
marker_distrib = marker_strategies(markers_symbol, marker_strategy, C)
```
## Deconvolution and evaluation for bulk RNA-seq methods
```{r}
fraction_out = data.frame()
evaluation_out = data.frame()
bulk_methods = c("CIBERSORT","nnls","FARDEEP")
for(i in bulk_methods){
    RESULTS = Deconvolution(T = T, C = C, phenoDataC = pDataC, method = i, P = P, elem = to_remove, marker_distrib = marker_distrib, refProfiles.var = refProfiles.var) 
    RESULTS$method = i
    fraction_out = rbind(fraction_out,RESULTS)
    evaluation = RESULTS %>% dplyr::summarise(RMSE = sqrt(mean((observed_values-expected_values)^2)) %>% round(.,4), 
                                              Pearson=cor(observed_values,expected_values) %>% round(.,4))
    evaluation$method = i
    evaluation_out = rbind(evaluation_out, evaluation)
}
```
## Deconvolution and evaluation for single-cell methods
```{r}
sc_methods = c("MuSiC","SCDC","BisqueRNA")
for(i in sc_methods){
    if(i == "BisqueRNA"){
      RESULTS = Deconvolution(T = T_orig, C = train_cellID, method = i, phenoDataC = pDataC, P = P, elem = to_remove, refProfiles.var = refProfiles.var)
    }else{
      RESULTS = Deconvolution(T = T, C = train_cellID_TMM, method = i, phenoDataC = pDataC, P = P, elem = to_remove, refProfiles.var = refProfiles.var) 
    }
    RESULTS$method = i
    fraction_out = rbind(fraction_out,RESULTS)
    evaluation = RESULTS %>% dplyr::summarise(RMSE = sqrt(mean((observed_values-expected_values)^2)) %>% round(.,4), 
                                              Pearson=cor(observed_values,expected_values) %>% round(.,4))
    evaluation$method = i
    evaluation_out = rbind(evaluation_out, evaluation)
}
write.table(fraction_out,file=paste0("Deconvolution_fraction.tsv"),sep="\t",quote = FALSE,row.names = FALSE,col.names = TRUE)
write.table(evaluation_out,file=paste0("Deconvolution_evaluation.tsv"),sep="\t",quote = FALSE,row.names = FALSE,col.names = TRUE)
write.table(sample_subset,file=paste0("Deconvolution_trainSamples.tsv"),sep="\t",quote = FALSE,row.names = FALSE,col.names = TRUE)
```
## Plot figures
```{r}
library(ggplot2)
library(tidyr)
library(dplyr)
```

### Prepare inpute files
```{r}
training = read.delim("Deconvolution_trainSamples.tsv")
fraction_raw = read.delim("Deconvolution_fraction.tsv")
fraction_raw = fraction_raw[fraction_raw$method != "FARDEEP" & fraction_raw$method != "nnls",]
fraction = fraction_raw %>% pivot_longer(c(observed_values,expected_values),names_to = "type",values_to = "fraction")
fraction_observed = fraction[fraction$type=="observed_values",]
fraction_expected = fraction[fraction$type=="expected_values",]
fraction_expected$method = "snRNA-seq"
fraction_all = rbind(fraction_observed,unique(fraction_expected))
fraction_all$method = gsub("BisqueRNA","Bisque", fraction_all$method)
```

### Performance of different deconvolution methods
```{r}
fraction_test = fraction_all[! fraction_all$tissue %in% training$x,]
p1 = ggplot(fraction_test,aes(x=CT,y=fraction,color=method))+geom_boxplot()+
  ylab("Fraction") +xlab("Cell Type")+
  theme_bw()+theme(panel.border = element_rect(colour = "black"),panel.grid.major = element_blank(), panel.grid.minor = element_blank(),axis.text=element_text(color="black",size=12),axis.title=element_text(color="black",size=13),legend.text=element_text(color="black",size=13),legend.title=element_text(color="black",size=13))
p1
ggsave(p1, file=paste0("Deconvolution_test.pdf"), width=7, height=4)
```
### Evaluation of different deconvolution methods
```{r}
fraction_test_eval = fraction_raw %>% dplyr::filter(! tissue %in% training$x) %>% group_by(tissue,method) %>% 
  summarise(RMSE = sqrt(mean((observed_values-expected_values)^2)) %>% round(.,4),Pearson=cor(observed_values,expected_values) %>% round(.,4))
fraction_test_eval_sum  = fraction_test_eval %>% group_by(method) %>% summarise(RMSE_mean = mean(RMSE),Pearson_mean = mean(Pearson))
```
```{r}
p2 = ggplot(fraction_test_eval,aes(x=method,y=Pearson,color=method))+geom_boxplot()+
  ylab("Correlation") +xlab("Method")+
  theme_bw()+theme(panel.border = element_rect(colour = "black"),panel.grid.major = element_blank(), panel.grid.minor = element_blank(),axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),axis.text=element_text(color="black",size=12),axis.title=element_text(color="black",size=13),legend.text=element_text(color="black",size=13),legend.title=element_text(color="black",size=13))
p2
ggsave(p2, file=paste0("Deconvolution_test.cor.pdf"), width=4, height=3)
```
```{r}
p3 = ggplot(fraction_test_eval,aes(x=method,y=RMSE,color=method))+geom_boxplot()+
  ylab("RMSE") +xlab("Method")+
  theme_bw()+theme(panel.border = element_rect(colour = "black"),panel.grid.major = element_blank(), panel.grid.minor = element_blank(),axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),axis.text=element_text(color="black",size=12),axis.title=element_text(color="black",size=13),legend.text=element_text(color="black",size=13),legend.title=element_text(color="black",size=13))
p3
ggsave(p3, file=paste0("Deconvolution_test.RMSE.pdf"), width=4, height=3)
```
## Session info
```{r}
sessionInfo()
```
